stackslib/src/net/httpcore.rs

```diff
@@ -0,0 +1,1625 @@
+                            "read http response payload of {} bytes (just buffered {})",
+                            message_bytes.len(),
+                            num_read,
+                        );
+
+                        // we now know the content-length, so pass it into the parser.
+                        let handler_index =
+                            self.request_handler_index
+                                .ok_or(NetError::DeserializeError(
+                                    "Unknown HTTP response handler".to_string(),
+                                ))?;
+
+                        let parse_res = self.try_parse_response(
+                            handler_index,
+                            http_response_preamble,
+                            &message_bytes[..],
+                        );
+
+                        // done parsing
+                        self.reset();
+                        match parse_res {
+                            Ok(data_response) => Ok((
+                                Some((
+                                    StacksHttpMessage::Response(data_response),
+                                    total_bytes_consumed,
+                                )),
+                                num_read,
+                            )),
+                            Err(e) => {
+                                info!("Failed to parse HTTP response: {:?}", &e);
+                                Err(e)
+                            }
+                        }
+                    }
+                    None => {
+                        // need more data
+                        trace!(
+                            "did not read http response payload, but buffered {}",
+                            num_read
+                        );
+                        Ok((None, num_read))
+                    }
+                }
+            }
+        }
+    }
+
+    /// Parse a payload of known length.
+    /// Only gets called if payload_len() returns Some(...).
+    ///
+    /// Return Ok(message, num-bytes-consumed) if we decoded a message.  The message will
+    /// have consumed `num-bytes-consumed` bytes.
+    ///
+    /// Return Err(..) if we failed to decode the message.
+    fn read_payload(
+        &mut self,
+        preamble: &StacksHttpPreamble,
+        buf: &[u8],
+    ) -> Result<(StacksHttpMessage, usize), NetError> {
+        match preamble {
+            StacksHttpPreamble::Request(ref http_request_preamble) => {
+                // all requests have a known length
+                let len = http_request_preamble.get_content_length() as usize;
+                assert!(len <= buf.len(), "{} > {}", len, buf.len());
+
+                trace!("read http request payload of {} bytes", len);
+
+                match self.try_parse_request(http_request_preamble, &buf[0..len]) {
+                    Ok(data_request) => Ok((StacksHttpMessage::Request(data_request), len)),
+                    Err(e) => {
```

Comment:
Simlarly to other error match comment, this double match can be reduced to a single match.

Reply:
Fixed